src/IB/IBFESurfaceMethod.cpp

```diff
@@ -640,48 +779,389 @@ IBFESurfaceMethod::interpolateVelocity(const int u_data_idx,
+
+                    for (unsigned int i = 0; i < NDIM; ++i)
+                    {
+                        for (unsigned int k = 0; k < n_node; ++k)
                         {
-                            x_qp[NDIM * (qp_offset + qp) + d] += x_node[k][d] * p;
+                            x_qp[NDIM * (qp_offset + qp) + i] += x_node[k][i] * phi[k][qp];
                         }
+                        N_qp[NDIM * (qp_offset + qp) + i] = n(i);
                     }
+
+                    for (unsigned int d = 0; d < NDIM; ++d)
+                        for (unsigned int i = 0; i < NDIM; ++i)
+                            for (unsigned int k = 0; k < n_node; ++k)
+                                DU_j_qp[d][NDIM * (qp_offset + qp) + i] += DU_j_node[d][k][i] * phi[k][qp];
                 }
                 qp_offset += n_qp;
             }
-
-            // Interpolate values from the Cartesian grid patch to the
-            // quadrature points.
-            //
-            // NOTE: Values are interpolated only to those quadrature points
-            // that are within the patch interior.
-            const Box<NDIM>& interp_box = patch->getBox();
-            Pointer<PatchData<NDIM> > u_data = patch->getPatchData(u_data_idx);
+            // Interpolate values from the Cartesian grid patch to the quadrature
+            // points.
+            // Note: Values are interpolated only to those quadrature points that
+            // are within the patch interior
+            const Box<NDIM>& patch_box = patch->getBox();
+            Pointer<PatchData<NDIM> > u_data = patch->getPatchData(side_mask_scratch_idx);
             Pointer<CellData<NDIM, double> > u_cc_data = u_data;
             if (u_cc_data)
             {
                 LEInteractor::interpolate(
-                    U_qp, NDIM, x_qp, NDIM, u_cc_data, patch, interp_box, d_default_interp_spec.kernel_fcn);
+                    U_qp, NDIM, x_qp, NDIM, u_cc_data, patch, patch_box, d_default_interp_spec.kernel_fcn);
             }
+
             Pointer<SideData<NDIM, double> > u_sc_data = u_data;
+
             if (u_sc_data)
             {
-                LEInteractor::interpolate(
-                    U_qp, NDIM, x_qp, NDIM, u_sc_data, patch, interp_box, d_default_interp_spec.kernel_fcn);
+                Pointer<SideData<NDIM, double> > u_sc_data = patch->getPatchData(side_mask_scratch_idx);
+
+                const IntVector<NDIM>& u_gcw = u_sc_data->getGhostCellWidth();
+
+                const int u_depth = u_sc_data->getDepth();
+                TBOX_ASSERT(u_depth == 1);
+
+                std::vector<int> local_indices;
+                local_indices.clear();
+                const unsigned int upper_bound = n_qp_patch;
+                if (upper_bound == 0) return;
+                local_indices.reserve(upper_bound);
+
+                for (unsigned int k = 0; k < upper_bound; ++k)
+                {
+                    const double* const XX = &x_qp[NDIM * k];
```

Comment:
Please don't mix uppercase X and lower case x --- they have different meanings. (X indicates reference coordinates, x indicates current coordinates.)